packages/web3-eth/src/eth_tx.ts

```diff
@@ -0,0 +1,315 @@
+
+import Web3Eth from '.';
+import {
+	ChainIdMismatchError,
+	CommonOrChainAndHardforkError,
+	Eip1559GasPriceError,
+	Eip1559NotSupportedError,
+	InvalidGasOrGasPrice,
+	InvalidMaxPriorityFeePerGasOrMaxFeePerGas,
+	InvalidNonceOrChainIdError,
+	InvalidTransactionObjectError,
+	MissingChainOrHardforkError,
+	MissingCustomChainError,
+	MissingCustomChainIdError,
+	MissingGasError,
+	UnableToPopulateNonceError,
+	UnsupportedFeeMarketError,
+	UnsupportedTransactionTypeError,
+} from './errors';
+import { PopulatedUnsignedTransaction, Transaction } from './types';
+
+export function formatTransaction<
+	DesiredType extends ValidTypes = ValidTypes,
+	ReturnType = ValidReturnTypes[DesiredType],
+>(
+	transaction: Transaction,
+	desiredType: DesiredType,
+	overrideMethod?: (transaction: Transaction) => Transaction<ReturnType>,
+): Transaction<ReturnType> {
+	if (overrideMethod !== undefined) return overrideMethod(transaction);
+	const formattedTransaction = {
+		...transaction,
+		value:
+			transaction.value === '0x' ? '0x' : convertToValidType(transaction.value, desiredType),
+		gas: convertToValidType(transaction.gas, desiredType),
+		gasPrice: convertToValidType(transaction.gasPrice, desiredType),
+		type: convertToValidType(transaction.type, desiredType),
+		maxFeePerGas: convertToValidType(transaction.maxFeePerGas, desiredType),
+		maxPriorityFeePerGas: convertToValidType(transaction.maxPriorityFeePerGas, desiredType),
+		nonce: convertToValidType(transaction.nonce, desiredType),
+		chainId: convertToValidType(transaction.chainId, desiredType),
+		gasLimit: convertToValidType(transaction.gasLimit, desiredType),
+		v: convertToValidType(transaction.v, desiredType),
+		common: {
+			...transaction.common,
+			customChain: {
+				...transaction.common?.customChain,
+				networkId: convertToValidType(
+					transaction.common?.customChain?.networkId,
+					desiredType,
+				),
+				chainId: convertToValidType(transaction.common?.customChain?.chainId, desiredType),
+			},
+		},
+	};
+	// TODO - TSC is complaining that ReturnType could be instantiated with an
+	// arbitrary type which could be unrelated to 'string | number | bigint | undefined'
+	return formattedTransaction as unknown as Transaction<ReturnType>;
+}
+
+export const detectTransactionType = (
+	transaction: Transaction,
+	overrideMethod?: (transaction: Transaction) => Numbers | undefined,
+): Numbers | undefined => {
+	if (overrideMethod !== undefined) return overrideMethod(transaction);
+	if (transaction.type !== undefined) return transaction.type;
+
+	if (transaction.maxFeePerGas !== undefined || transaction.maxPriorityFeePerGas !== undefined)
+		return '0x2';
+	if (transaction.hardfork === 'london') return '0x2';
+	if (transaction.common?.hardfork === 'london') return '0x2';
+
+	if (transaction.accessList !== undefined) return '0x1';
+	if (transaction.hardfork === 'berlin') return '0x1';
+	if (transaction.common?.hardfork === 'berlin') return '0x1';
+
+	return undefined;
```

Comment:
If tx is legacy transaction without an envelope, it should be type 0.

Reply:
I would like to keep the `return undefined` so that it's explicit when this method fails to detect type (throwing an error is also an option, users would then just wrap in a `try/catch`). However, we could add detection for `0x0` transaction that most input to this method will probably satisfy

Reply:
The [1.x implementation](https://github.com/ChainSafe/web3.js/blob/8783f4d64e424456bdc53b34ef1142d0a7cee4d7/packages/web3-eth-accounts/src/index.js#L353) (`_handleTxType`) would return whatever was passed via `transaction.type` or `undefined`, so how `detectTransactionType` runs now matches `1.x`

Reply:
This clean implementation looks more good as compared to 1.x, In 4.x we should not mutate obj / update tx type inside this function and should return tx type. 

The changes in tx object should be done inside one `filling defaults function` ( unless user explicitly specified not to via boolean )